nattable.rb

```diff
@@ -164,19 +164,33 @@ def empty_port(remote_addr, remote_port)
   end
 
   def local_key_from_packet(packet)
-    packet.tuple + packet.l4.tuple
+    l3_tuple = packet.tuple
+    l3_tuple_size = l3_tuple.size
+
+    b = IO::Buffer.new(l3_tuple_size + 4)
+    b.copy(l3_tuple)
+    b.copy(packet.l4.tuple, l3_tuple_size)
+
+    b.get_string
```

Comment:
Probably, the reason we are not seeing speedup with NAT is due to these lines. Honestly, I doubt if this function and `remote_key_from_packet` have become slower by swithing to IO::Buffer.

Here, we are building the key to lookup the NAT table, by concatenating the IP address tuple and the port tuple.

It could be the case that the String class of ruby has optimizations for handling tiny strings as well as concatenating them, while IO::Buffer does not have something alike.

Also, `IO::Buffer` cannot be used as a hash key and we have to call `IO::Buffer#get_string`. That can be costing us as well.

Reply:
> Also, `IO::Buffer` cannot be used as a hash key and we have to call `IO::Buffer#get_string`. That can be costing us as well.

Maybe IO::Buffer should implement `hash` and `eql?` based on the byte contents? cc @ioquatix

Reply:
We could do that but for a hash key it can be tricky, since it can be mutated.

Reply:
A String can be mutated too, but Hash actually does .dup.freeze when putting a non-frozen String into the Hash.
I'm not sure mutation is an issue, but indeed it would be awkward for IO::Buffer to cache the hash, maybe there is no need to cache it in the IO::Buffer though.

Anyway, this seems a good case for `transfer_string`.

Reply:
I think one could use `IO::Buffer.for` here to workaround the lack of `transfer_string`:
```ruby
    l3_tuple = packet.tuple
    l3_tuple_size = l3_tuple.size
    string = "\0".b * (l3_tuple_size + 4) # "\0".b could be stored in a constant

    IO::Buffer.for(string) do |b|
      b.copy(l3_tuple)
      b.copy(packet.l4.tuple, l3_tuple_size)
    end

    string
```
This should avoid the extra bytes copy.
Of course there is still an allocation of `l3_tuple_size + 4` bytes but there were 2 of them in the code above.

I think this is less elegant/more convoluted than `.transfer_string` (and it may be hard to support on some Ruby implementations without an extra copy on `.for`) but it's one way that already works now.

Reply:
Thank you for the suggestion! Applied in 7dca4c3.

Ideally, I would prefer doing `(packet.tuple + packet.l4.tuple).transfer_string`, because then there would be zero intermediary objects and all the lengths and offsets can be calculated in C.

But this is better than what I wrote.